File: sgRNA_count.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(ShortRead)
  library(parallel)
  library(stringr)
  library(dplyr)
  library(optparse)
})

# Define command line options
option_list <- list(
  make_option(c("-f", "--fastq"), type="character", default=NULL,
              help="Comma-separated list of FASTQ files", metavar="character"),
  make_option(c("-n", "--names"), type="character", default=NULL,
              help="Comma-separated sample names corresponding to FASTQ files", metavar="character"),
  make_option(c("-c", "--ncpu"), type="integer", default=10,
              help="Number of CPU cores to use [default= %default]", metavar="integer"),
  make_option(c("-l", "--library"), type="character", default=NULL,
              help="sgRNA library CSV file", metavar="character"),
  make_option(c("-o", "--output"), type="character", default="sgRNA_counts",
              help="Output prefix [default= %default]", metavar="character")
)

# Parse arguments
opt_parser <- OptionParser(option_list=option_list)
opt <- parse_args(opt_parser)

# Check required arguments
if (is.null(opt$fastq) | is.null(opt$library)) {
  print_help(opt_parser)
  stop("Both --fastq and --library arguments must be supplied", call.=FALSE)
}

# Record start time
start_time <- Sys.time()

# 1. Load sgRNA library
message("Loading sgRNA library: ", opt$library)
gRNA_library <- read.csv(opt$library, stringsAsFactors = FALSE)
target_sequences <- unique(gRNA_library$gRNA.sequence)

# 2. Process input files and sample names
fastq_files <- unlist(strsplit(opt$fastq, ","))

if (!is.null(opt$names)) {
  sample_names <- unlist(strsplit(opt$names, ","))
  if (length(fastq_files) != length(sample_names)) {
    stop("Number of FASTQ files must match number of sample names", call.=FALSE)
  }
} else {
  # Generate sample names from filenames if not provided
  sample_names <- tools::file_path_sans_ext(basename(fastq_files))
  sample_names <- gsub("_R1\\.fq\\.gz$", "", sample_names)
}

# 3. Initialize final result table
final_result <- gRNA_library %>%
  select(id, gRNA.sequence, Gene)

# 4. Process each FASTQ file
for (i in seq_along(fastq_files)) {
  fastq_file <- fastq_files[i]
  sample_name <- sample_names[i]
  
  message("\nProcessing sample ", i, " of ", length(fastq_files), ": ", sample_name)
  message("  FASTQ file: ", fastq_file)
  
  # Read FASTQ
  message("  Reading FASTQ file...")
  fq <- readFastq(fastq_file)
  reads <- as.character(sread(fq))
  
  # Parallel counting
  message("  Counting sgRNA matches using ", opt$ncpu, " CPU cores...")
  cl <- makeCluster(opt$ncpu, type = "PSOCK")
  clusterExport(cl, varlist = c("reads", "str_detect", "fixed"), envir = environment())
  
  count_list <- parLapply(cl, target_sequences, function(seq) {
    sum(str_detect(reads, fixed(seq)))
  })
  
  stopCluster(cl)
  
  # Create result data frame
  result <- data.frame(
    gRNA.sequence = target_sequences,
    count = unlist(count_list)
  )
  
  # Add to final result
  final_result <- final_result %>%
    left_join(result, by = "gRNA.sequence") %>%
    rename(!!sample_name := count)
  
  # Save individual result
  individual_result <- gRNA_library %>%
    left_join(result, by = "gRNA.sequence") %>%
    mutate(count = ifelse(is.na(count), 0, count)) %>%
    select(id, gRNA.sequence, Gene, count) %>%
    rename(!!sample_name := count)
  
  indiv_file <- paste0(opt$output, "_", sample_name, ".csv")
  message("  Writing individual results to: ", indiv_file)
  write.csv(individual_result, indiv_file, row.names = FALSE)
}

# 5. Final processing and output
message("\nFinalizing results...")
final_result[is.na(final_result)] <- 0

merged_file <- paste0(opt$output, "_merged_all_samples.csv")
message("Writing merged results to: ", merged_file)
write.csv(final_result, merged_file, row.names = FALSE)

# 6. Print runtime
end_time <- Sys.time()
runtime <- end_time - start_time
message("\nAnalysis completed!")
message("Total runtime: ", round(runtime, 2), " ", units(runtime))

File: sgRNA_count.py
#!/usr/bin/env python3
import gzip
import argparse
from multiprocessing import Pool
import pandas as pd
from Bio import SeqIO
from tqdm import tqdm
import os

def parse_args():
    parser = argparse.ArgumentParser(description='Count sgRNA sequences in FASTQ files')
    parser.add_argument('-f', '--fastq', required=True,
                        help='Comma-separated list of FASTQ files')
    parser.add_argument('-l', '--library', required=True,
                        help='sgRNA library CSV file')
    parser.add_argument('-n', '--names', default=None,
                        help='Comma-separated sample names')
    parser.add_argument('-c', '--ncpu', type=int, default=10,
                        help='Number of CPU cores to use')
    parser.add_argument('-o', '--output', default='sgRNA_counts',
                        help='Output prefix')
    return parser.parse_args()

def load_library(library_file):
    return pd.read_csv(library_file)

def count_sequences(args):
    seq, fastq_file = args
    count = 0
    with gzip.open(fastq_file, 'rt') as f:
        for record in SeqIO.parse(f, 'fastq'):
            if seq in str(record.seq):
                count += 1
    return (seq, count)

def process_sample(fastq_file, library, ncpu):
    targets = library['gRNA.sequence'].unique()
    
    with Pool(ncpu) as pool:
        results = list(tqdm(pool.imap(count_sequences, 
                                    [(seq, fastq_file) for seq in targets]),
                      total=len(targets),
                      desc=f"Processing {os.path.basename(fastq_file)}"))
    
    return pd.DataFrame(results, columns=['gRNA.sequence', 'count'])

def main():
    args = parse_args()
    
    # Load library
    library = load_library(args.library)
    
    # Process samples
    fastq_files = args.fastq.split(',')
    sample_names = args.names.split(',') if args.names else [
        os.path.basename(f).replace('_R1.fq.gz', '').replace('.fq.gz', '') 
        for f in fastq_files
    ]
    
    merged_counts = library.copy()
    
    for fastq_file, name in zip(fastq_files, sample_names):
        print(f"\nProcessing {name} ({fastq_file})")
        counts = process_sample(fastq_file, library, args.ncpu)
        merged_counts = merged_counts.merge(
            counts, on='gRNA.sequence', how='left'
        ).rename(columns={'count': name})
        
        # Save individual results
        individual = library.merge(counts, on='gRNA.sequence', how='left')
        individual[name] = individual['count'].fillna(0)
        individual[['id', 'gRNA.sequence', 'Gene', name]].to_csv(
            f"{args.output}_{name}.csv", index=False
        )
    
    # Save merged results
    merged_counts.fillna(0, inplace=True)
    merged_counts.to_csv(f"{args.output}_merged_all_samples.csv", index=False)
    print("\nAnalysis completed!")

if __name__ == '__main__':
    main()
